src/levenshtein.cr

```diff
@@ -13,36 +13,68 @@ module Levenshtein
   def self.distance(string1 : String, string2 : String) : Int32
     return 0 if string1 == string2
 
-    s = string1.chars
-    t = string2.chars
-
-    s_size = s.size
-    t_size = t.size
+    s_size = string1.size
+    t_size = string2.size
 
     return t_size if s_size == 0
     return s_size if t_size == 0
 
-    # This is to allocate less memory
-    if t_size > s_size
-      t, s = s, t
-      t_size, s_size = s_size, t_size
-    end
+    if string1.ascii_only? && string2.ascii_only?
+      s = string1.to_unsafe
+      t = string2.to_unsafe
 
-    v = Pointer(Int32).malloc(t_size + 1) { |i| i }
+      # This is to allocate less memory
+      if t_size > s_size
+        t, s = s, t
+        t_size, s_size = s_size, t_size
+      end
 
-    s_size.times do |i|
-      last_cost = i + 1
+      costs = Pointer(Int32).malloc(t_size + 1) { |i| i }
 
-      t_size.times do |j|
-        sub_cost = s[i] == t[j] ? 0 : 1
-        cost = Math.min(Math.min(last_cost + 1, v[j + 1] + 1), v[j] + sub_cost)
-        v[j] = last_cost
-        last_cost = cost
+      last_cost = 0
+      s_size.times do |i|
+        last_cost = i + 1
+
+        t_size.times do |j|
+          sub_cost = s[i] == t[j] ? 0 : 1
+          cost = Math.min(Math.min(last_cost + 1, costs[j + 1] + 1), costs[j] + sub_cost)
+          costs[j] = last_cost
+          last_cost = cost
+        end
+        costs[t_size] = last_cost
       end
-      v[t_size] = last_cost
-    end
 
-    v[t_size]
+      last_cost
+    else
+      reader = Char::Reader.new(string1)
+
+      # Use an array instead of a reader to decode the second string only once
+      chars = string2.chars
+
+      # This is to allocate less memory
+      if t_size > s_size
+        chars, reader = reader, chars
+        t_size = s_size
+      end
```

Comment:
Shouldn't we make sure to call `chars` on the shorter string only?

Reply:
Are you suggesting to change line 52 to this?
```cr
chars = if t_size > s_size
  string1.chars
else
  string2.chars
end
```

Reply:
I think that makes sense. Need to switch the reader too, though. 
It would probably make sense to switch `string1` and `string2` right at the beginning, after determining the size.

Reply:
> I think that makes sense. Need to switch the reader too, though.

Then I suppose it would become this:
```cr
if t_size > s_size
  chars = string1.chars
  reader = Char::Reader.new(string2)
else
  chars = string2.chars
  reader = Char::Reader.new(string1)
end
```
> It would probably make sense to switch `string1` and `string2` right at the beginning, after determining the size. So that `string1` is always the shorter one.

So `string1` takes the bytes of `string2` and `string2` takes the bytes of `string1`. How would you imagine this to be implemented? Something using `Pointer#swap`?

Reply:
just
```cr
if t_size > s_size
  string1, string2 = string2, string1
end
```